### PR TITLE
Use forked gocheck

### PR DESCRIPTION
**Pull request by cpuguy83**
- added support for per-check timeouts (upstream is non-responsive)
- set 5m per-check timeout

---

The original PR I made to go-check had a bad rebase (and difficult to recover from), so opened a new on here: go-check/check#80

Another PR for consideration is go-check/check#70 which adds parallel tests -- this is not included here as I have not merged it into my fork's master branch in consideration of the per-check timeout being a bit more pertinent to Docker's test suite right now.

The last merge to go-check/check was 3 months ago, and that merge also broke the test suite :(

### CODE REVIEW

**Comment by cpuguy83** [comment]
ping @tiborvass @icecrime @tianon

**Comment by jfrazelle** [comment]
ah saddness that upstream :(

**Comment by thaJeztah** [comment]
I see there's also an open PR from @brahmaroutu https://github.com/go-check/check/pull/47 should we include that in our fork?

**Comment by tianon** [comment]
Oh Go ecosystem :cry:

For posterity's sake, can we enumerate some of the issues we've seen with upstream here?  (perhaps links to PRs that have gone stale?)  That'll be helpful to point anyone asking why we use a fork (or who traces it back to a commit and thus PR), and could also be an OK way to check whether we could consider switching back sometime later.

**Comment by cpuguy83** [comment]
@tianon Yes, can definitely do that. Unfortunately I'll have to re-create the branch since I managed to `push -f` a change on it that really should have gone in another branch ( :cry: ). Luckily the changes required for this are rather small other than adding the tests.

**Comment by thaJeztah** [comment]
Related PR upstream: https://github.com/go-check/check/pull/70 (and the one I mentioned above possibly)

**Comment by tianon** [comment]
@cpuguy83 oh, I just meant here -- it's pretty easy IMO to trace a commit back to a PR (by going to https://github.com/.../commit/... and clicking the `#xxx` link), so @thaJeztah's comment is good enough for me :+1:

**Comment by cpuguy83** [comment]
Updated the PR to list the PRs.

**Comment by tianon** [comment]
LGTM :disappointed:

**Comment by icecrime** [comment]
Not sure what's wrong here:

```
15:27:32 can't load package: package github.com/docker/docker: no buildable Go source files in /go/src/github.com/docker/docker
```

**Comment by thaJeztah** [comment]
ping @cpuguy83 any idea what those failures are? TL;DR got a lot of WindowsTP4 builds timing out, so wondering if we should up the timeout for those, or if we can get this PR merged soon

**Comment by cpuguy83** [comment]
I have literally no idea.
Note that experimental also failed just like janky, even though it is marked as successful.

**Comment by thaJeztah** [comment]
ping @mlaventure perhaps we should up the timeout for WindowsTP4 while we figure out what's wrong here?

**Comment by mlaventure** [comment]
This one is a timeout issue.

It looks like the copy of the sources into the container didn't happen for some reason.

But I can see it happening in the log:
```
04:26:34 Step 50 : COPY . /go/src/github.com/docker/docker
04:26:38  ---> ba1e4201c70f
```

**Comment by cpuguy83** [comment]
Yeah, it's super strange and it's only happening on this PR.
Tried rebuilding multiple times... rebasing and pushing, etc.

**Comment by cpuguy83** [comment]
And win2lin and userns run successfully.

**Comment by cpuguy83** [comment]
Seems I found the issue. Had to move the `check.timeout` flag from make.sh to make/integration.
Tests seem to be making it past where they did before.

**Comment by calavera** [comment]
LGTM

**Comment by cyphar** [comment]
If this works well, I think we should ditch using the suite timeout at all (it doesn't make sense, because we're always adding new tests, it's more important that if a test stalls we fail it after a timeout -- it doesn't matter how long the whole suite takes as long as it doesn't freeze).

**Comment by cpuguy83** [comment]
@cyphar Really I believe the `-timeout` option would be per-test, but since we're hijacking a single testing.T in the integration tests it ends up being for the whole suite.

Instead of removing it, we can still set the timeout option so it also applies to unit tests and such.

File: vendor/src/github.com/go-check/check/README.md [hunk 1]
```diff
@@ -1,20 +1,10 @@
...
 
-Install the package with:
+This is a fork of https://github.com/go-check/check
 
-    go get gopkg.in/check.v1
-    
-Import it with:
+The intention of this fork is not to change any of the original behavior, but add
+some specific behaviors needed for some of my projects already using this test suite.
+For documentation on the main behavior of go-check see the afformentioned repo.
```
Comment (tiborvass) [inline vendor/src/github.com/go-check/check/README.md]: `aforementioned` :)

File: hack/make.sh [hunk 2]
```diff
@@ -170,10 +170,13 @@ BUILDFLAGS=( $BUILDFLAGS "${ORIG_BUILDFLAGS[@]}" )
...
+	: ${CHECKTIMEOUT:=10m}
 elif [ "${DOCKER_ENGINE_GOARCH}" == "windows" ]; then
 	: ${TIMEOUT:=180m}
+	: ${CHECKTIMEOUT:=8m}
 else
 	: ${TIMEOUT:=120m}
+	: ${CHECKTIMEOUT:=5m}
 fi
 
 TESTFLAGS+=" -test.timeout=${TIMEOUT}"
```
Comment (tiborvass) [inline hack/make.sh]: Would like to remove this, and add `-test.timout=${TIMEOUT}` manually to hack/make/test-unit. That way only unit tests use this. This would also allow for removing the CHECKTIMEOUT variable, and simply using the TIMEOUT for what it shoudl be used: per-test timeout. It's just that unit tests and integration tests are implemented differently.

File: Makefile [hunk 3]
```diff
@@ -45,7 +45,7 @@ DOCKER_ENVS := \
 	-e DOCKER_USERLANDPROXY \
 	-e TESTDIRS \
 	-e TESTFLAGS \
-	-e TIMEOUT
+	-e TIMEOUT \
```
Comment (estesp) [inline Makefile]: not going to harm anything at the moment, but this line continuation isn't needed after removing the new variable

**Comment by cpuguy83** [comment]
all :green_heart: again
In order to remove the somewhat duplicated `CHECKTIMEOUT` (and use `TIMEOUT` instead), we had to set a super long timeout for the integration tests.
This is because there is no way to disable go's built-in test timeout, which defaults to calling `SIGQUIT` after 10mins (when `-timeout` is not set)

File: hack/make.sh [hunk 4]
```diff
@@ -250,6 +248,7 @@ test_env() {
 		DOCKER_REMAP_ROOT="$DOCKER_REMAP_ROOT" \
 		DOCKER_REMOTE_DAEMON="$DOCKER_REMOTE_DAEMON" \
 		GOPATH="$GOPATH" \
+		GOTRACEBACK=all \
```
Comment (cpuguy83) [inline hack/make.sh]: This was added so that the panics that go-check produces give a full goroutine dump, otherwise the trace is useless since it's just a trace of the main go-check code, not the timed-out test code.

File: hack/make/test-integration-cli [hunk 5]
```diff
@@ -2,7 +2,7 @@
 set -e
 
 bundle_test_integration_cli() {
-	TESTFLAGS="$TESTFLAGS -check.v"
+	TESTFLAGS="$TESTFLAGS -check.v -check.timeout=${TIMEOUT} -timeout=360m"
```
Comment (tianon) [inline hack/make/test-integration-cli]: Does just `-timeout` here work, or does it have to be `-test.timeout` explicitly?

Comment (tianon) [inline hack/make/test-integration-cli]: (I'm guessing by the green that it's fine, but just triple-checking.)

**Comment by tianon** [comment]
LGTM

File: hack/make/test-integration-cli (hunk 5 above)
Comment (cpuguy83) [inline hack/make/test-integration-cli]: yeah, none of the built-in flags need the `test` prefix (or at least none that I've used).

**Comment by tiborvass** [comment]
LGTM

**Comment by thaJeztah** [comment]
:tada: thanks @cpuguy83
